helix-term/src/application.rs

```diff
@@ -241,64 +248,128 @@ impl Application {
+                            }
+                            (None, Some(message), None) => {
+                                format!("[{}] {}", token_d, message)
+                            }
+                            (None, None, Some(percentage)) => {
+                                format!("[{}] {}%", token_d, percentage)
+                            }
+                            (None, None, None) => format!("[{}]", token_d),
                         };
-                        let status = format!("[{}] {}", token, msg);
+
+                        if let lsp::WorkDoneProgress::End(_) = work {
+                            self.lsp_progress.end_progress(server_id, &token);
+                        } else {
+                            self.lsp_progress.update(server_id, token, work);
+                        }
+
                         self.editor.set_status(status);
                         self.render();
                     }
                     _ => unreachable!(),
                 }
             }
-            Call::MethodCall(call) => {
-                error!("Method not found {}", call.method);
+            Call::MethodCall(helix_lsp::jsonrpc::MethodCall {
+                method,
+                params,
+                jsonrpc,
+                id,
+            }) => {
+                let call = match MethodCall::parse(&method, params) {
+                    Some(call) => call,
+                    None => {
+                        error!("Method not found {}", method);
+                        return;
+                    }
+                };
 
+                match call {
+                    MethodCall::WorkDoneProgressCreate(params) => {
+                        self.lsp_progress.create(server_id, params.token);
+
+                        let doc = self.editor.documents().find(|doc| {
+                            doc.language_server()
+                                .map(|server| server.id() == server_id)
+                                .unwrap_or_default()
+                        });
+                        match doc {
+                            Some(doc) => {
+                                // it's ok to unwrap, we check for the language server before
+                                let server = doc.language_server().unwrap();
+                                helix_lsp::block_on(server.reply(id, Ok(serde_json::Value::Null)));
+                            }
+                            None => {
+                                if let Some(server) =
+                                    self.editor.language_servers.get_by_id(server_id)
+                                {
+                                    log::warn!(
+                                        "missing document with language server id `{}`",
+                                        server_id
+                                    );
+                                    helix_lsp::block_on(server.reply(
```

Comment:
Since this method is async, you can use `server.reply().await`

Reply:
Ideally though, we want to use `tokio::spawn` here to avoid blocking. The way to do that is to change `reply` to look more like notify:

https://github.com/helix-editor/helix/blob/1f2f97051632550da0decb6f646cc2904804b2fc/helix-lsp/src/client.rs#L150-L203

Instead of being `async fn`, which borrows the `server` for an indefinite amount of time until completion (which is why rustc asks for a static lifetime) we construct an `async move` closure and return it as a Future.  I updated both `call` and `notify` in this way, but nothing was using `reply` yet so it still has the old form.

Reply:
Interesting, I somehow missed that `handle_language_server_message` is async. I also tried changing the `Client::reply` to reflect `Client::notify` but it didn't work. Now when I tried it though it works.